File: dsa_algorithm.py
"""
Distribution Structure Analysis (DSA) Algorithm
Python Implementation
License: MIT
Author: Michio Okazaki
Contact: senryaku@si-lab.work
"""

import numpy as np
import pandas as pd
from scipy import stats
from scipy.optimize import minimize
import warnings
from datetime import datetime
from typing import Dict, List, Tuple, Optional

warnings.filterwarnings('ignore')


class DSAAlgorithm:
    """
    Distribution Structure Analysis (DSA) Algorithm
    
    A comprehensive framework for rigorous and transparent statistical analysis
    in medical research with automated audit trail generation.
    """
    
    def __init__(self, data: np.ndarray, estimand: str = "Mean difference between groups",
                 alpha: float = 0.05):
        """
        Initialize DSA Algorithm
        
        Parameters:
        -----------
        data : np.ndarray
            Numeric array of observations
        estimand : str
            Description of the estimand
        alpha : float
            Significance level for statistical tests (default: 0.05)
        """
        self.data = np.array(data)
        self.estimand = estimand
        self.alpha = alpha
        self.audit_log = {
            'timestamp': datetime.now(),
            'estimand': estimand,
            'sample_size': len(data),
            'data_summary': self._get_data_summary(),
            'distributions_tested': []
        }
        
    def _get_data_summary(self) -> Dict:
        """Get summary statistics of the data"""
        return {
            'mean': np.mean(self.data),
            'median': np.median(self.data),
            'std': np.std(self.data),
            'min': np.min(self.data),
            'max': np.max(self.data),
            'q25': np.percentile(self.data, 25),
            'q75': np.percentile(self.data, 75)
        }
    
    def run(self, distributions: Optional[List[str]] = None) -> Dict:
        """
        Run the DSA algorithm
        
        Parameters:
        -----------
        distributions : List[str], optional
            List of distributions to test. If None, tests all available distributions.
            
        Returns:
        --------
        Dict containing selected distribution, parameters, GOF metrics, and audit log
        """
        if distributions is None:
            distributions = ['normal', 'lognormal', 'exponential', 'weibull', 
                           'gamma', 'poisson', 'negbinom']
        
        print("=== DSA Algorithm Started ===")
        print(f"Estimand: {self.estimand}")
        print(f"Sample size: {len(self.data)}\n")
        
        # Step 1: Data validation
        if np.any(np.isnan(self.data)):
            raise ValueError("Data contains missing values. Please handle missing data before analysis.")
        
        # Step 2: Distribution identification
        print("Step 2: Testing candidate distributions...")
        results = self._test_distributions(distributions)
        
        # Step 3: Goodness-of-fit assessment
        print("\nStep 3: Goodness-of-fit assessment...")
        gof_results = self._assess_goodness_of_fit(results)
        
        # Step 4: Select best distribution
        print("\nStep 4: Selecting best distribution...")
        best_dist = self._select_best_distribution(gof_results)
        
        # Step 5: Quality control flags
        qc_flags = self._assign_quality_flags(best_dist)
        
        # Update audit log
        self.audit_log['results'] = results
        self.audit_log['gof_results'] = gof_results
        self.audit_log['selected_distribution'] = best_dist
        self.audit_log['qc_flags'] = qc_flags
        self.audit_log['completion_time'] = datetime.now()
        
        # Print summary
        print("\n=== DSA Algorithm Completed ===")
        print(f"Selected distribution: {best_dist['distribution']}")
        print(f"Quality flag: {qc_flags['flag']}")
        print(f"AIC: {best_dist['aic']:.2f}")
        print(f"BIC: {best_dist['bic']:.2f}\n")
        
        return {
            'distribution': best_dist['distribution'],
            'parameters': best_dist['parameters'],
            'gof_metrics': best_dist['gof_metrics'],
            'qc_flags': qc_flags,
            'audit_log': self.audit_log
        }
    
    def _test_distributions(self, distributions: List[str]) -> Dict:
        """Test multiple candidate distributions"""
        results = {}
        
        for dist in distributions:
            try:
                if dist == 'normal':
                    params = stats.norm.fit(self.data)
                    loglik = np.sum(stats.norm.logpdf(self.data, *params))
                    n_params = 2
                    
                elif dist == 'lognormal':
                    if np.any(self.data <= 0):
                        print(f"  Skipping lognormal (data contains non-positive values)")
                        continue
                    params = stats.lognorm.fit(self.data)
                    loglik = np.sum(stats.lognorm.logpdf(self.data, *params))
                    n_params = 3
                    
                elif dist == 'exponential':
                    if np.any(self.data < 0):
                        print(f"  Skipping exponential (data contains negative values)")
                        continue
                    params = stats.expon.fit(self.data)
                    loglik = np.sum(stats.expon.logpdf(self.data, *params))
                    n_params = 2
                    
                elif dist == 'weibull':
                    if np.any(self.data < 0):
                        print(f"  Skipping Weibull (data contains negative values)")
                        continue
                    params = stats.weibull_min.fit(self.data)
                    loglik = np.sum(stats.weibull_min.logpdf(self.data, *params))
                    n_params = 3
                    
                elif dist == 'gamma':
                    if np.any(self.data <= 0):
                        print(f"  Skipping gamma (data contains non-positive values)")
                        continue
                    params = stats.gamma.fit(self.data)
                    loglik = np.sum(stats.gamma.logpdf(self.data, *params))
                    n_params = 3
                    
                elif dist == 'poisson':
                    if np.any(self.data < 0) or not np.all(self.data == np.floor(self.data)):
                        print(f"  Skipping Poisson (data must be non-negative integers)")
                        continue
                    lambda_param = np.mean(self.data)
                    params = (lambda_param,)
                    loglik = np.sum(stats.poisson.logpmf(self.data.astype(int), lambda_param))
                    n_params = 1
                    
                elif dist == 'negbinom':
                    if np.any(self.data < 0) or not np.all(self.data == np.floor(self.data)):
                        print(f"  Skipping negative binomial (data must be non-negative integers)")
                        continue
                    # Fit negative binomial using method of moments
                    mean = np.mean(self.data)
                    var = np.var(self.data)
                    if var <= mean:
                        print(f"  Skipping negative binomial (variance <= mean)")
                        continue
                    n = mean**2 / (var - mean)
                    p = mean / var
                    params = (n, p)
                    loglik = np.sum(stats.nbinom.logpmf(self.data.astype(int), n, p))
                    n_params = 2
                    
                else:
                    print(f"  Unknown distribution: {dist}")
                    continue
                
                # Calculate AIC and BIC
                n = len(self.data)
                aic = 2 * n_params - 2 * loglik
                bic = n_params * np.log(n) - 2 * loglik
                
                results[dist] = {
                    'params': params,
                    'aic': aic,
                    'bic': bic,
                    'loglik': loglik,
                    'n_params': n_params
                }
                
                print(f"  {dist}: AIC = {aic:.2f}, BIC = {bic:.2f}")
                
            except Exception as e:
                print(f"  Error fitting {dist}: {str(e)}")
        
        self.audit_log['distributions_tested'] = list(results.keys())
        return results
    
    def _assess_goodness_of_fit(self, results: Dict) -> Dict:
        """Assess goodness-of-fit for all candidate distributions"""
        gof_results = {}
        
        for dist_name, result in results.items():
            # Kolmogorov-Smirnov test
            if dist_name == 'normal':
                ks_stat, ks_pvalue = stats.kstest(self.data, 'norm', args=result['params'])
            elif dist_name == 'lognormal':
                ks_stat, ks_pvalue = stats.kstest(self.data, 'lognorm', args=result['params'])
            elif dist_name == 'exponential':
                ks_stat, ks_pvalue = stats.kstest(self.data, 'expon', args=result['params'])
            elif dist_name == 'weibull':
                ks_stat, ks_pvalue = stats.kstest(self.data, 'weibull_min', args=result['params'])
            elif dist_name == 'gamma':
                ks_stat, ks_pvalue = stats.kstest(self.data, 'gamma', args=result['params'])
            else:
                ks_stat, ks_pvalue = np.nan, np.nan
            
            gof_results[dist_name] = {
                'aic': result['aic'],
                'bic': result['bic'],
                'ks_statistic': ks_stat,
                'ks_pvalue': ks_pvalue,
                'params': result['params']
            }
        
        return gof_results
    
    def _select_best_distribution(self, gof_results: Dict) -> Dict:
        """Select best distribution based on hierarchical criteria"""
        # Priority 3: Information criteria (BIC)
        bic_values = {name: result['bic'] for name, result in gof_results.items()}
        best_dist_name = min(bic_values, key=bic_values.get)
        
        return {
            'distribution': best_dist_name,
            'parameters': gof_results[best_dist_name]['params'],
            'aic': gof_results[best_dist_name]['aic'],
            'bic': gof_results[best_dist_name]['bic'],
            'gof_metrics': gof_results[best_dist_name]
        }
    
    def _assign_quality_flags(self, best_dist: Dict) -> Dict:
        """Assign quality control flags (Red/Yellow/Green)"""
        ks_pvalue = best_dist['gof_metrics']['ks_pvalue']
        
        if np.isnan(ks_pvalue):
            flag = "YELLOW"
            message = "K-S test not applicable for discrete distributions"
        elif ks_pvalue < 0.01:
            flag = "RED"
            message = "Poor fit: Consider alternative distributions or transformations"
        elif ks_pvalue < 0.05:
            flag = "YELLOW"
            message = "Marginal fit: Proceed with caution and sensitivity analysis"
        else:
            flag = "GREEN"
            message = "Good fit: Distribution is appropriate"
        
        return {'flag': flag, 'message': message, 'ks_pvalue': ks_pvalue}
    
    def generate_audit_report(self, output_file: str = "dsa_audit_report.txt"):
        """Generate audit report"""
        with open(output_file, 'w') as f:
            f.write("=== DSA Algorithm Audit Report ===\n\n")
            f.write(f"Timestamp: {self.audit_log['timestamp']}\n")
            f.write(f"Estimand: {self.audit_log['estimand']}\n")
            f.write(f"Sample size: {self.audit_log['sample_size']}\n\n")
            
            f.write("Data Summary:\n")
            for key, value in self.audit_log['data_summary'].items():
                f.write(f"  {key}: {value:.4f}\n")
            f.write("\n")
            
            f.write(f"Distributions Tested: {', '.join(self.audit_log['distributions_tested'])}\n\n")
            
            if 'selected_distribution' in self.audit_log:
                f.write(f"Selected Distribution: {self.audit_log['selected_distribution']['distribution']}\n")
                f.write(f"Quality Flag: {self.audit_log['qc_flags']['flag']}\n")
                f.write(f"Quality Message: {self.audit_log['qc_flags']['message']}\n\n")
            
            f.write(f"Completion Time: {self.audit_log['completion_time']}\n")
        
        print(f"Audit report saved to: {output_file}")


# Example usage
if __name__ == "__main__":
    # Generate example data
    np.random.seed(123)
    data = np.random.normal(50, 10, 1000)
    
    # Run DSA algorithm
    dsa = DSAAlgorithm(data, estimand="Mean treatment effect")
    result = dsa.run()
    
    # Generate audit report
    dsa.generate_audit_report()

File: dsa_algorithm.R
# Distribution Structure Analysis (DSA) Algorithm
# R Implementation
# License: MIT
# Author: Michio Okazaki
# Contact: senryaku@si-lab.work

# Required packages
required_packages <- c("fitdistrplus", "MASS", "mixtools", "ggplot2", "dplyr")

# Install missing packages
install_if_missing <- function(packages) {
  for (pkg in packages) {
    if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
      install.packages(pkg, dependencies = TRUE)
      library(pkg, character.only = TRUE)
    }
  }
}

install_if_missing(required_packages)

#' DSA Algorithm: Main Function
#' 
#' @param data Numeric vector of observations
#' @param estimand Character string describing the estimand
#' @param distributions Vector of distribution names to test (default: all)
#' @param alpha Significance level for statistical tests (default: 0.05)
#' @return List containing selected distribution, parameters, GOF metrics, and audit log
#' @export
dsa_algorithm <- function(data, 
                          estimand = "Mean difference between groups",
                          distributions = c("normal", "lognormal", "exponential", 
                                           "weibull", "gamma", "poisson", "negbin"),
                          alpha = 0.05) {
  
  # Initialize audit log
  audit_log <- list(
    timestamp = Sys.time(),
    estimand = estimand,
    sample_size = length(data),
    data_summary = summary(data),
    distributions_tested = distributions
  )
  
  cat("=== DSA Algorithm Started ===\n")
  cat("Estimand:", estimand, "\n")
  cat("Sample size:", length(data), "\n\n")
  
  # Step 1: Data validation
  if (any(is.na(data))) {
    stop("Data contains missing values. Please handle missing data before analysis.")
  }
  
  # Step 2: Distribution identification
  cat("Step 2: Testing candidate distributions...\n")
  results <- test_distributions(data, distributions, alpha)
  
  # Step 3: Goodness-of-fit assessment
  cat("\nStep 3: Goodness-of-fit assessment...\n")
  gof_results <- assess_goodness_of_fit(data, results)
  
  # Step 4: Select best distribution
  cat("\nStep 4: Selecting best distribution...\n")
  best_dist <- select_best_distribution(gof_results)
  
  # Step 5: Quality control flags
  qc_flags <- assign_quality_flags(best_dist, gof_results)
  
  # Update audit log
  audit_log$results <- results
  audit_log$gof_results <- gof_results
  audit_log$selected_distribution <- best_dist
  audit_log$qc_flags <- qc_flags
  audit_log$completion_time <- Sys.time()
  
  # Print summary
  cat("\n=== DSA Algorithm Completed ===\n")
  cat("Selected distribution:", best_dist$distribution, "\n")
  cat("Quality flag:", qc_flags$flag, "\n")
  cat("AIC:", round(best_dist$aic, 2), "\n")
  cat("BIC:", round(best_dist$bic, 2), "\n\n")
  
  return(list(
    distribution = best_dist$distribution,
    parameters = best_dist$parameters,
    gof_metrics = best_dist$gof_metrics,
    qc_flags = qc_flags,
    audit_log = audit_log
  ))
}

#' Test multiple candidate distributions
#' @keywords internal
test_distributions <- function(data, distributions, alpha) {
  results <- list()
  
  for (dist in distributions) {
    tryCatch({
      if (dist == "normal") {
        fit <- fitdist(data, "norm")
      } else if (dist == "lognormal") {
        if (any(data <= 0)) {
          cat("  Skipping lognormal (data contains non-positive values)\n")
          next
        }
        fit <- fitdist(data, "lnorm")
      } else if (dist == "exponential") {
        if (any(data < 0)) {
          cat("  Skipping exponential (data contains negative values)\n")
          next
        }
        fit <- fitdist(data, "exp")
      } else if (dist == "weibull") {
        if (any(data < 0)) {
          cat("  Skipping Weibull (data contains negative values)\n")
          next
        }
        fit <- fitdist(data, "weibull")
      } else if (dist == "gamma") {
        if (any(data <= 0)) {
          cat("  Skipping gamma (data contains non-positive values)\n")
          next
        }
        fit <- fitdist(data, "gamma")
      } else if (dist == "poisson") {
        if (any(data < 0) || any(data != floor(data))) {
          cat("  Skipping Poisson (data must be non-negative integers)\n")
          next
        }
        fit <- fitdist(data, "pois", discrete = TRUE)
      } else if (dist == "negbin") {
        if (any(data < 0) || any(data != floor(data))) {
          cat("  Skipping negative binomial (data must be non-negative integers)\n")
          next
        }
        fit <- fitdist(data, "nbinom", discrete = TRUE)
      } else {
        cat("  Unknown distribution:", dist, "\n")
        next
      }
      
      results[[dist]] <- list(
        fit = fit,
        aic = fit$aic,
        bic = fit$bic,
        loglik = fit$loglik
      )
      
      cat("  ", dist, ": AIC =", round(fit$aic, 2), ", BIC =", round(fit$bic, 2), "\n")
      
    }, error = function(e) {
      cat("  Error fitting", dist, ":", e$message, "\n")
    })
  }
  
  return(results)
}

#' Assess goodness-of-fit for all candidate distributions
#' @keywords internal
assess_goodness_of_fit <- function(data, results) {
  gof_results <- list()
  
  for (dist_name in names(results)) {
    fit <- results[[dist_name]]$fit
    
    # Kolmogorov-Smirnov test
    ks_test <- gofstat(fit, fitnames = dist_name)
    
    gof_results[[dist_name]] <- list(
      aic = results[[dist_name]]$aic,
      bic = results[[dist_name]]$bic,
      ks_statistic = ks_test$ks,
      ks_pvalue = ks_test$kstest
    )
  }
  
  return(gof_results)
}

#' Select best distribution based on hierarchical criteria
#' @keywords internal
select_best_distribution <- function(gof_results) {
  # Priority 1: Theoretical plausibility (handled by user)
  # Priority 2: Visual diagnostics (Q-Q plot, P-P plot)
  # Priority 3: Information criteria (AIC/BIC)
  # Priority 4: Statistical tests (K-S test)
  
  # Select distribution with lowest BIC
  bic_values <- sapply(gof_results, function(x) x$bic)
  best_dist_name <- names(which.min(bic_values))
  
  return(list(
    distribution = best_dist_name,
    parameters = gof_results[[best_dist_name]],
    aic = gof_results[[best_dist_name]]$aic,
    bic = gof_results[[best_dist_name]]$bic,
    gof_metrics = gof_results[[best_dist_name]]
  ))
}

#' Assign quality control flags (Red/Yellow/Green)
#' @keywords internal
assign_quality_flags <- function(best_dist, gof_results) {
  ks_pvalue <- best_dist$gof_metrics$ks_pvalue
  
  if (ks_pvalue < 0.01) {
    flag <- "RED"
    message <- "Poor fit: Consider alternative distributions or transformations"
  } else if (ks_pvalue < 0.05) {
    flag <- "YELLOW"
    message <- "Marginal fit: Proceed with caution and sensitivity analysis"
  } else {
    flag <- "GREEN"
    message <- "Good fit: Distribution is appropriate"
  }
  
  return(list(flag = flag, message = message, ks_pvalue = ks_pvalue))
}

#' Generate Q-Q plot for visual diagnostics
#' @export
plot_qq <- function(data, distribution, parameters) {
  qqplot_data <- qqplot(data, distribution, parameters)
  print(qqplot_data)
}

#' Generate audit report
#' @export
generate_audit_report <- function(audit_log, output_file = "dsa_audit_report.txt") {
  sink(output_file)
  
  cat("=== DSA Algorithm Audit Report ===\n\n")
  cat("Timestamp:", as.character(audit_log$timestamp), "\n")
  cat("Estimand:", audit_log$estimand, "\n")
  cat("Sample size:", audit_log$sample_size, "\n\n")
  
  cat("Data Summary:\n")
  print(audit_log$data_summary)
  cat("\n")
  
  cat("Distributions Tested:", paste(audit_log$distributions_tested, collapse = ", "), "\n\n")
  
  cat("Selected Distribution:", audit_log$selected_distribution$distribution, "\n")
  cat("Quality Flag:", audit_log$qc_flags$flag, "\n")
  cat("Quality Message:", audit_log$qc_flags$message, "\n\n")
  
  cat("Completion Time:", as.character(audit_log$completion_time), "\n")
  
  sink()
  
  cat("Audit report saved to:", output_file, "\n")
}

# Example usage (commented out)
# set.seed(123)
# data <- rnorm(1000, mean = 50, sd = 10)
# result <- dsa_algorithm(data, estimand = "Mean treatment effect")
# generate_audit_report(result$audit_log)
